### ISSUE TITLE
Multiple session handler options

### ISSUE THREAD

**Issue by phoenixgao** [bolt/bolt#4499]
If someone would like to work on multiple session handler options, (like redis), would it be an extension or included in base?

Or, does an extension have this ability to change default session handler(storage)?

**Comment by GawainLynch** [comment]
Already in master

**Comment by phoenixgao** [comment]
awesome!

**Comment by GawainLynch** [comment]
I thought you might like that :grin:

**Comment by phoenixgao** [comment]
I can't figure out how to use it, tried to fix it but got confused by some code:

1. save_handler is hard coded to "files", no way to config it: but it's fixable
https://github.com/bolt/bolt/blob/master/src/Provider/SessionServiceProvider.php#L223

2. $connetions here are hard coded to null, also not reading from config
https://github.com/bolt/bolt/blob/master/src/Provider/SessionServiceProvider.php#L400

3. save_path is fixed to /tmp so if connections is not configured this save_path will mess up the return value 
https://github.com/bolt/bolt/blob/master/src/Provider/SessionServiceProvider.php#L400
so defaults here https://github.com/bolt/bolt/blob/master/src/Provider/SessionServiceProvider.php#L384
and for memcached are useless
4. $connections seems an array contains multiple connection schema (host, port, etc..)but in this foreach... it just override and override by last one:
https://github.com/bolt/bolt/blob/master/src/Provider/SessionServiceProvider.php#L351-L363

same for memcache handler.



I'm not sure how to fix it properly. Also it checks phpredis first then predis. It could be nicer if user can choose(config) which redis driver to use.

**Comment by GawainLynch** [comment]
The master branch is still a work in progress… it is not ready for production, or even beta, use yet.

As I pointed to, the parts are there, but some things are dependant on other work still being done.

**Comment by SahAssar** [comment]
I'm pretty sure this will be better tested when we have a "fuller" implementation of it, and we can have issues for specific bugs.

I'm tagging this for closing on the 11:th unless good reason otherwise, and I look forward to seeing issues/PR's for this functionality when we have a proper base to test on :smile:

**Comment by GawainLynch** [comment]
@SahAssar agreed… the feed back is very useful @phoenixgao, it is just that we have shifted focus in the last week from master becoming v2.3 to making it v3.0 and that has created extra work.

@SahAssar It might, in immediate hindsight, be worth keeping this open as a tracker on the problem as @phoenixgao is a willing victim… I mean tester… of the functionality! Does that work for you, @phoenixgao ?

**Comment by phoenixgao** [comment]
Oh because I always work on project which master branch is always a stable and fully function one... if your master branch is a work in progress then it's fine :D 
Sorry for not being able to work on bolt recently.

Feel free to close this issue or keep it open as a tracker~

I'm a big fan of bolt, it makes me want to write blogs ;P while wordpress didn't work for me, for years... I've never set up wordpress successfully, with git

**Comment by SahAssar** [comment]
@GawainLynch Leaving this open per your suggestion.

@phoenixgao master is our development branch, it's bound to be unstable at certain times... Especially when the :koala: is around ;)

**Comment by ToBe998** [comment]
This issue just went critical on me. It is currently not possible to use Bolt3 in any kind of cluster setup.
The memcache/redis/whatever options can not beused as the config is overridden and it allways uses filesystem sessions.
You also can not just share the sessions folder between servers as the code does not allowe the use of symlinks here.

Is there any news on this issue? This basically stops all upgrading projects and new sites for me atm.

**Comment by GawainLynch** [comment]
Erm, didn't your needed changes go into `release/3.1`?

**Comment by ToBe998** [comment]
I didnt hear any news on the release plans for 3.1 so i assumed it's to planned in the next time?
If it is, that would be fine.

**Comment by ToBe998** [comment]
Just a small heads up, im still having session problems, btu they may not be related to this issue. While debugging however, i found out that `memcache` is supported now, but `memcached` has a small bug. The code returns the proper service for memcached, but uses the wrong Pimple identifier, leading to this error:

```
Error: Uncaught exception 'InvalidArgumentException' with message 'Identifier "session.handler_factory.backing_memcached" is not defined.' in /vendor/pimple/pimple/lib/Pimple.php:78
```

Nothing imported since memcach "should" be a good substitute, but still maybe unexpected.

**Comment by SahAssar** [comment]
memcached/memcache should work now I think (although I have not used them personally), could you confirm/deny that @ToBe998 ?

**Comment by ToBe998** [comment]
Im running **memcached** with config settings in my config_local.yml without trouble for some time now. The other options (auto prsed options from php.ini or memcache did not run back then, did not test since. But for me it's ok for now.

**Comment by SahAssar** [comment]
@ToBe998: I'm very late on this, but would you be willing to contribute some docs for that or (if you don't have the time to write the docs), share a config example here and then I can perhaps write something up?

**Comment by ToBe998** [comment]
Sure. I just shied away from that git driven doc system until now. I'll give it a try. ;)

**Comment by bobdenotter** [comment]
@ToBe998 Even I managed to set it up, using the [README](https://github.com/bolt/docs/blob/site/README.md). And if you're only making changes to a single page, the "quick PR" functionality in Github is pretty neat. :-)

**Comment by ToBe998** [comment]
Damn. You just took away my last excuse to not do it. :D

**Comment by luispabon** [comment]
Ping. Impossible to use Bolt on Kubernetes (autoscaling, pod upgrades...) or any other load balanced scenario due to lack of redis/memcached session handling.

**Comment by luispabon** [comment]
Workaround: configure php itself to store sessions in redis. Cumbersome, but works:

```ini
session.save_handler = redis
session.save_path = "tcp://redis:6379"
```

**Comment by GawainLynch** [comment]
What version of Bolt were you doing that one BTW?

**Comment by luispabon** [comment]
You asking me? Latest, although above would work with anything that defers sessions to php session handling.

**Comment by bobdenotter** [comment]
> Latest,

Please be more specific. Which version? Installed how?

**Comment by GawainLynch** [comment]
> You asking me? 

Oui, the branches are in huge flux and a tonne has changed in that area between 3.2.8+ on an ongoing basis :smile:

**Comment by GawainLynch** [comment]
This should be implemented in the just released 3.3.0. If there are relevant problems please open a new issue with details.
